// @milkdown/crepe/lib/cjs/feature/ai/index.js
'use strict';

var diff$1 = require('@milkdown/kit/component/diff');
var diff = require('@milkdown/kit/plugin/diff');
var streaming = require('@milkdown/kit/plugin/streaming');
var ctx = require('@milkdown/kit/ctx');
var core = require('@milkdown/kit/core');
var exception = require('@milkdown/kit/exception');
var utils = require('@milkdown/kit/utils');
var state = require('@milkdown/kit/prose/state');
var DOMPurify = require('dompurify');
var tooltip = require('@milkdown/kit/plugin/tooltip');
var prose = require('@milkdown/kit/prose');
var vue = require('vue');
var component = require('@milkdown/kit/component');
var view = require('@milkdown/kit/prose/view');

ctx.createSlice([], "FeaturesCtx");
ctx.createSlice({}, "CrepeCtx");
function useCrepeFeatures(ctx) {
  return ctx.use("FeaturesCtx");
}
function crepeFeatureConfig(feature) {
  return (ctx) => {
    useCrepeFeatures(ctx).update((features) => {
      if (features.includes(feature)) {
        return features;
      }
      return [...features, feature];
    });
  };
}

var CrepeFeature = /* @__PURE__ */ ((CrepeFeature2) => {
  CrepeFeature2["CodeMirror"] = "code-mirror";
  CrepeFeature2["ListItem"] = "list-item";
  CrepeFeature2["LinkTooltip"] = "link-tooltip";
  CrepeFeature2["Cursor"] = "cursor";
  CrepeFeature2["ImageBlock"] = "image-block";
  CrepeFeature2["BlockEdit"] = "block-edit";
  CrepeFeature2["Toolbar"] = "toolbar";
  CrepeFeature2["Placeholder"] = "placeholder";
  CrepeFeature2["Table"] = "table";
  CrepeFeature2["Latex"] = "latex";
  CrepeFeature2["TopBar"] = "top-bar";
  CrepeFeature2["AI"] = "ai";
  return CrepeFeature2;
})(CrepeFeature || {});

function defaultBuildContext(ctx, instruction) {
  var _a;
  const view = ctx.get(core.editorViewCtx);
  const serializer = ctx.get(core.serializerCtx);
  const { state } = view;
  const document = serializer(state.doc);
  let selection = "";
  if (!state.selection.empty) {
    const { from, to } = state.selection;
    const slice = state.doc.slice(from, to);
    const { schema } = state.doc.type;
    let wrapper = schema.topNodeType.createAndFill(null, slice.content);
    if (!wrapper) {
      const paragraph = (_a = schema.nodes.paragraph) == null ? void 0 : _a.createAndFill(
        null,
        slice.content
      );
      if (paragraph) wrapper = schema.topNodeType.createAndFill(null, paragraph);
    }
    selection = wrapper ? serializer(wrapper) : state.doc.textBetween(from, to);
  }
  return { document, selection, instruction };
}

const aiProviderConfig = utils.$ctx(
  {
    provider: void 0,
    buildContext: void 0,
    diffReviewOnEnd: true,
    onError: (error) => {
      console.error(`[milkdown/ai] [${error.code}]`, error);
    },
    aiIcon: void 0
  },
  "aiProviderConfig"
);
const aiSessionCtx = utils.$ctx(
  {
    abortController: null,
    label: "",
    lastInstruction: "",
    lastLabel: void 0,
    lastFrom: -1,
    lastTo: -1,
    diffOwnedByAI: false
  },
  "aiSession"
);
function emitAIError(ctx, error) {
  const config = ctx.get(aiProviderConfig.key);
  try {
    config.onError(error);
  } catch (handlerError) {
    console.error("[milkdown/ai] onError handler failed:", handlerError);
  }
}
function clearActiveSession(ctx) {
  const current = ctx.get(aiSessionCtx.key);
  ctx.set(aiSessionCtx.key, {
    ...current,
    abortController: null,
    label: ""
  });
}
async function runProvider(ctx, provider, promptContext, abortController) {
  try {
    const iterable = provider(promptContext, abortController.signal);
    const commands = ctx.get(core.commandsCtx);
    for await (const chunk of iterable) {
      if (abortController.signal.aborted) break;
      commands.call(streaming.pushChunkCmd.key, chunk);
    }
    if (abortController.signal.aborted) return;
    const config = ctx.get(aiProviderConfig.key);
    if (config.diffReviewOnEnd) {
      const cur = ctx.get(aiSessionCtx.key);
      ctx.set(aiSessionCtx.key, { ...cur, diffOwnedByAI: true });
    }
    const dispatched = commands.call(streaming.endStreamingCmd.key, {
      diffReview: config.diffReviewOnEnd
    });
    if (config.diffReviewOnEnd && !dispatched) {
      const cur = ctx.get(aiSessionCtx.key);
      ctx.set(aiSessionCtx.key, { ...cur, diffOwnedByAI: false });
    }
  } catch (error) {
    if (abortController.signal.aborted) return;
    const milkdownError = exception.aiProviderError(error);
    emitAIError(ctx, milkdownError);
    const commands = ctx.get(core.commandsCtx);
    commands.call(streaming.abortStreamingCmd.key, { keep: false });
  } finally {
    const current = ctx.get(aiSessionCtx.key);
    if (current.abortController === abortController) {
      clearActiveSession(ctx);
    }
  }
}
const runAICmd = utils.$command("RunAI", (ctx) => {
  return (options) => (state, dispatch) => {
    var _a, _b, _c, _d;
    if (!(options == null ? void 0 : options.instruction)) return false;
    const config = ctx.get(aiProviderConfig.key);
    if (!config.provider) return false;
    const session = ctx.get(aiSessionCtx.key);
    if (session.abortController) return false;
    if ((_a = streaming.streamingPluginKey.getState(state)) == null ? void 0 : _a.active) return false;
    if ((_b = diff.diffPluginKey.getState(state)) == null ? void 0 : _b.active) return false;
    if (!dispatch) return true;
    const abortController = new AbortController();
    const { from, to } = state.selection;
    ctx.set(aiSessionCtx.key, {
      abortController,
      label: (_c = options.label) != null ? _c : "",
      lastInstruction: options.instruction,
      lastLabel: options.label,
      lastFrom: from,
      lastTo: to,
      // Reset every run; only the success path that hands off to diff
      // review flips it back on.
      diffOwnedByAI: false
    });
    const commands = ctx.get(core.commandsCtx);
    const insertAt = state.selection.empty ? "cursor" : "selection";
    if (!commands.call(streaming.startStreamingCmd.key, { insertAt })) {
      clearActiveSession(ctx);
      return false;
    }
    let promptContext;
    try {
      const buildContext = (_d = config.buildContext) != null ? _d : defaultBuildContext;
      promptContext = buildContext(ctx, options.instruction);
    } catch (error) {
      const milkdownError = exception.aiBuildContextError(error);
      emitAIError(ctx, milkdownError);
      commands.call(streaming.abortStreamingCmd.key, { keep: false });
      clearActiveSession(ctx);
      return false;
    }
    void runProvider(ctx, config.provider, promptContext, abortController);
    return true;
  };
});
const abortAICmd = utils.$command("AbortAI", (ctx) => {
  return (options) => (state, dispatch) => {
    var _a;
    const session = ctx.get(aiSessionCtx.key);
    if (!dispatch) return !!session.abortController;
    if (!session.abortController) return false;
    session.abortController.abort();
    clearActiveSession(ctx);
    if ((_a = streaming.streamingPluginKey.getState(state)) == null ? void 0 : _a.active) {
      const commands = ctx.get(core.commandsCtx);
      commands.call(streaming.abortStreamingCmd.key, options);
    }
    return true;
  };
});

const aiIcon = `
  <svg
    xmlns="http://www.w3.org/2000/svg"
    width="24"
    height="24"
    viewBox="0 0 24 24"
  >
    <path
      fill="currentColor"
      d="M19 9l1.25-2.75L23 5l-2.75-1.25L19 1l-1.25 2.75L15 5l2.75 1.25zm0 14l1.25-2.75L23 19l-2.75-1.25L19 15l-1.25 2.75L15 19l2.75 1.25zM9 20l3-6.5L18 10l-6-3.5L9 0L6 6.5L0 10l6 3.5z"
      transform="translate(3 3) scale(0.75)"
    />
  </svg>
`;

const chevronLeftIcon = `
  <svg
    xmlns="http://www.w3.org/2000/svg"
    fill="none"
    viewBox="0 0 24 24"
    stroke-width="1.5"
    stroke="currentColor"
  >
    <path
      stroke-linecap="round"
      stroke-linejoin="round"
      d="M15.75 19.5L8.25 12l7.5-7.5"
    />
  </svg>
`;

const chevronRightIcon = `
  <svg
    xmlns="http://www.w3.org/2000/svg"
    fill="none"
    viewBox="0 0 24 24"
    stroke-width="1.5"
    stroke="currentColor"
  >
    <path
      stroke-linecap="round"
      stroke-linejoin="round"
      d="M8.25 4.5l7.5 7.5-7.5 7.5"
    />
  </svg>
`;

const clearIcon = `
  <svg
    xmlns="http://www.w3.org/2000/svg"
    width="24"
    height="24"
    viewBox="0 0 24 24"
  >
    <g clip-path="url(#clip0_1098_15553)">
      <path
        d="M18.3007 5.70973C17.9107 5.31973 17.2807 5.31973 16.8907 5.70973L12.0007 10.5897L7.1107 5.69973C6.7207 5.30973 6.0907 5.30973 5.7007 5.69973C5.3107 6.08973 5.3107 6.71973 5.7007 7.10973L10.5907 11.9997L5.7007 16.8897C5.3107 17.2797 5.3107 17.9097 5.7007 18.2997C6.0907 18.6897 6.7207 18.6897 7.1107 18.2997L12.0007 13.4097L16.8907 18.2997C17.2807 18.6897 17.9107 18.6897 18.3007 18.2997C18.6907 17.9097 18.6907 17.2797 18.3007 16.8897L13.4107 11.9997L18.3007 7.10973C18.6807 6.72973 18.6807 6.08973 18.3007 5.70973Z"
      />
    </g>
    <defs>
      <clipPath id="clip0_1098_15553">
        <rect width="24" height="24" />
      </clipPath>
    </defs>
  </svg>
`;

const confirmIcon = `
  <svg
    xmlns="http://www.w3.org/2000/svg"
    width="24"
    height="24"
    viewBox="0 0 24 24"
    fill="none"
  >
    <g clip-path="url(#clip0_1013_1606)">
      <path
        d="M9.00012 16.1998L5.50012 12.6998C5.11012 12.3098 4.49012 12.3098 4.10012 12.6998C3.71012 13.0898 3.71012 13.7098 4.10012 14.0998L8.29012 18.2898C8.68012 18.6798 9.31012 18.6798 9.70012 18.2898L20.3001 7.69982C20.6901 7.30982 20.6901 6.68982 20.3001 6.29982C19.9101 5.90982 19.2901 5.90982 18.9001 6.29982L9.00012 16.1998Z"
        fill="#817567"
      />
    </g>
    <defs>
      <clipPath id="clip0_1013_1606">
        <rect width="24" height="24" />
      </clipPath>
    </defs>
  </svg>
`;

const editIcon = `
  <svg
    xmlns="http://www.w3.org/2000/svg"
    width="24"
    height="24"
    viewBox="0 0 24 24"
  >
    <g clip-path="url(#clip0_1013_1585)">
      <path
        d="M14.06 9.02L14.98 9.94L5.92 19H5V18.08L14.06 9.02ZM17.66 3C17.41 3 17.15 3.1 16.96 3.29L15.13 5.12L18.88 8.87L20.71 7.04C21.1 6.65 21.1 6.02 20.71 5.63L18.37 3.29C18.17 3.09 17.92 3 17.66 3ZM14.06 6.19L3 17.25V21H6.75L17.81 9.94L14.06 6.19Z"
      />
    </g>
    <defs>
      <clipPath id="clip0_1013_1585">
        <rect width="24" height="24" />
      </clipPath>
    </defs>
  </svg>
`;

const enterKeyIcon = `
  <svg
    xmlns="http://www.w3.org/2000/svg"
    width="24"
    height="24"
    viewBox="0 0 24 24"
  >
    <path
      fill="currentColor"
      d="M9 16l-5-5l5-5l1.4 1.4L7.825 10H18V6h2v6H7.825l2.575 2.6z"
    />
  </svg>
`;

const grammarCheckIcon = `
  <svg
    xmlns="http://www.w3.org/2000/svg"
    width="24"
    height="24"
    viewBox="0 0 24 24"
  >
    <path
      fill="currentColor"
      d="M9.55 17.575L4.225 12.25l1.4-1.425L9.55 14.75l8.825-8.825l1.4 1.425z"
    />
  </svg>
`;

const longerIcon = `
  <svg
    xmlns="http://www.w3.org/2000/svg"
    width="24"
    height="24"
    viewBox="0 0 24 24"
  >
    <path
      fill="currentColor"
      d="M3 18v-2h13v2zm0-5v-2h18v2zm0-5V6h18v2zm15 13v-3h-3v-2h3v-3h2v3h3v2h-3v3z"
    />
  </svg>
`;

const retryIcon = `
  <svg
    xmlns="http://www.w3.org/2000/svg"
    width="24"
    height="24"
    viewBox="0 0 24 24"
  >
    <path
      fill="currentColor"
      d="M12 20q-3.35 0-5.675-2.325T4 12q0-3.35 2.325-5.675T12 4q1.725 0 3.3.713T18 6.75V4h2v7h-7V9h4.2q-.8-1.4-2.187-2.2T12 6Q9.5 6 7.75 7.75T6 12q0 2.5 1.75 4.25T12 18q1.925 0 3.475-1.1T17.65 14h2.1q-.7 2.65-2.85 4.325T12 20"
    />
  </svg>
`;

const sendIcon = `
  <svg
    xmlns="http://www.w3.org/2000/svg"
    width="24"
    height="24"
    viewBox="0 0 24 24"
  >
    <path
      fill="currentColor"
      d="M11 20V7.825l-5.6 5.6L4 12l8-8l8 8l-1.4 1.425l-5.6-5.6V20z"
    />
  </svg>
`;

const sendPromptIcon = `
  <svg
    xmlns="http://www.w3.org/2000/svg"
    width="24"
    height="24"
    viewBox="0 0 24 24"
  >
    <path
      fill="currentColor"
      d="M3 20v-6l8-2l-8-2V4l19 8z"
    />
  </svg>
`;

const shorterIcon = `
  <svg
    xmlns="http://www.w3.org/2000/svg"
    width="24"
    height="24"
    viewBox="0 0 24 24"
  >
    <path
      fill="currentColor"
      d="M5 13v-2h14v2z"
    />
  </svg>
`;

const translateIcon = `
  <svg
    xmlns="http://www.w3.org/2000/svg"
    width="24"
    height="24"
    viewBox="0 0 24 24"
  >
    <path
      fill="currentColor"
      d="m12.9 16l-2.55-2.5l.05-.05q1.45-1.6 2.262-3.117T13.95 7H16V5h-6V3H8v2H2v2h9.1q-.5 1.2-1.225 2.387T8 11.7q-.6-.6-1.187-1.412T5.85 8.6H3.85q.5 1.4 1.3 2.7T6.95 13.7L2.85 17.75L4.25 19.15L8 15.4L10.55 17.95zM18.5 22h-2L21 10h2l4.5 12h-2l-1.05-3h-4.9zm1.6-5h3.6l-1.8-5.1z"
    />
  </svg>
`;

var __typeError$3 = (msg) => {
  throw TypeError(msg);
};
var __accessCheck$3 = (obj, member, msg) => member.has(obj) || __typeError$3("Cannot " + msg);
var __privateGet$3 = (obj, member, getter) => (__accessCheck$3(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateAdd$3 = (obj, member, value) => member.has(obj) ? __typeError$3("Cannot add the same private member more than once") : member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
var __privateSet$3 = (obj, member, value, setter) => (__accessCheck$3(obj, member, "write to private field"), member.set(obj, value), value);
var __privateMethod = (obj, member, method) => (__accessCheck$3(obj, member, "access private method"), method);
var _panel, _host, _retryBtn, _config, _visible, _diffActive, _diffStartDoc, _ownedByAI, _DiffActionsPanelView_instances, findHost_fn, makeButton_fn, makeShortcutChip_fn, _retry, canRetry_fn, _rejectAll, _acceptAll;
const PANEL_CLASS = "milkdown-ai-diff-actions";
function setSanitizedIcon(host, svg) {
  host.innerHTML = DOMPurify.sanitize(svg.trim());
}
function createIcon(svg) {
  const span = document.createElement("span");
  span.className = `${PANEL_CLASS}-icon`;
  setSanitizedIcon(span, svg);
  return span;
}
class DiffActionsPanelView {
  constructor(ctx, view, config) {
    this.ctx = ctx;
    __privateAdd$3(this, _DiffActionsPanelView_instances);
    __privateAdd$3(this, _panel);
    __privateAdd$3(this, _host);
    __privateAdd$3(this, _retryBtn);
    __privateAdd$3(this, _config);
    __privateAdd$3(this, _visible, false);
    /// Tracks the diff plugin's `active` flag across transactions so we
    /// can detect false→true / true→false edges independently of whether
    /// the panel is actually being shown.
    __privateAdd$3(this, _diffActive, false);
    /// Doc snapshot at the moment diff review activated. If the live doc
    /// drifts from this snapshot the user has accepted some per-change
    /// diffs and the stored `lastFrom`/`lastTo` no longer point at the
    /// original range — Retry is unsafe at that point.
    __privateAdd$3(this, _diffStartDoc, null);
    /// Whether the active diff review came from this AI session's
    /// streaming hand-off (vs being started manually via
    /// `startDiffReviewCmd`). Captured at the false→true transition.
    /// The panel only renders when this is true so it doesn't take over
    /// non-AI diff flows that exist independently of the AI feature.
    __privateAdd$3(this, _ownedByAI, false);
    __privateAdd$3(this, _retry, () => {
      const session = this.ctx.get(aiSessionCtx.key);
      if (!session.lastInstruction) return;
      if (!__privateGet$3(this, _ownedByAI) || !__privateMethod(this, _DiffActionsPanelView_instances, canRetry_fn).call(this)) return;
      const commands = this.ctx.get(core.commandsCtx);
      commands.call(diff.clearDiffReviewCmd.key);
      const editorView = this.ctx.get(core.editorViewCtx);
      const { doc } = editorView.state;
      const from = Math.min(Math.max(session.lastFrom, 0), doc.content.size);
      const to = Math.min(Math.max(session.lastTo, 0), doc.content.size);
      editorView.dispatch(
        editorView.state.tr.setSelection(
          state.TextSelection.create(editorView.state.doc, from, to)
        )
      );
      commands.call(runAICmd.key, {
        instruction: session.lastInstruction,
        label: session.lastLabel
      });
    });
    __privateAdd$3(this, _rejectAll, () => {
      this.ctx.get(core.commandsCtx).call(diff.clearDiffReviewCmd.key);
    });
    __privateAdd$3(this, _acceptAll, () => {
      this.ctx.get(core.commandsCtx).call(diff.acceptAllDiffsCmd.key);
    });
    __privateSet$3(this, _config, config);
    __privateSet$3(this, _host, __privateMethod(this, _DiffActionsPanelView_instances, findHost_fn).call(this, view));
    const panel = document.createElement("div");
    panel.className = PANEL_CLASS;
    panel.dataset.show = "false";
    __privateSet$3(this, _retryBtn, __privateMethod(this, _DiffActionsPanelView_instances, makeButton_fn).call(this, "retry", config.retryIcon, config.retryLabel, __privateGet$3(this, _retry)));
    panel.appendChild(__privateGet$3(this, _retryBtn));
    panel.appendChild(
      __privateMethod(this, _DiffActionsPanelView_instances, makeButton_fn).call(this, "reject", config.rejectIcon, config.rejectAllLabel, __privateGet$3(this, _rejectAll))
    );
    const acceptBtn = __privateMethod(this, _DiffActionsPanelView_instances, makeButton_fn).call(this, "accept", config.acceptIcon, config.acceptAllLabel, __privateGet$3(this, _acceptAll));
    acceptBtn.appendChild(__privateMethod(this, _DiffActionsPanelView_instances, makeShortcutChip_fn).call(this));
    panel.appendChild(acceptBtn);
    __privateSet$3(this, _panel, panel);
    __privateGet$3(this, _host).appendChild(panel);
    this.update(view);
  }
  update(view) {
    var _a;
    const diffActive = !!((_a = diff.diffPluginKey.getState(view.state)) == null ? void 0 : _a.active);
    if (diffActive !== __privateGet$3(this, _diffActive)) {
      __privateSet$3(this, _diffActive, diffActive);
      if (diffActive) {
        const session = this.ctx.get(aiSessionCtx.key);
        __privateSet$3(this, _ownedByAI, session.diffOwnedByAI);
        __privateSet$3(this, _diffStartDoc, view.state.doc);
      } else {
        __privateSet$3(this, _ownedByAI, false);
        __privateSet$3(this, _diffStartDoc, null);
        const session = this.ctx.get(aiSessionCtx.key);
        if (session.diffOwnedByAI) {
          this.ctx.set(aiSessionCtx.key, {
            ...session,
            diffOwnedByAI: false
          });
        }
      }
    }
    const shouldShow = diffActive && __privateGet$3(this, _ownedByAI);
    if (shouldShow !== __privateGet$3(this, _visible)) {
      __privateSet$3(this, _visible, shouldShow);
      __privateGet$3(this, _panel).dataset.show = shouldShow ? "true" : "false";
    }
    if (shouldShow) {
      const session = this.ctx.get(aiSessionCtx.key);
      const docUntouched = !!__privateGet$3(this, _diffStartDoc) && view.state.doc.eq(__privateGet$3(this, _diffStartDoc));
      __privateGet$3(this, _retryBtn).disabled = !session.lastInstruction || !docUntouched;
    }
  }
  destroy() {
    __privateGet$3(this, _panel).remove();
  }
}
_panel = new WeakMap();
_host = new WeakMap();
_retryBtn = new WeakMap();
_config = new WeakMap();
_visible = new WeakMap();
_diffActive = new WeakMap();
_diffStartDoc = new WeakMap();
_ownedByAI = new WeakMap();
_DiffActionsPanelView_instances = new WeakSet();
findHost_fn = function(view) {
  var _a;
  return (_a = view.dom.closest(".milkdown")) != null ? _a : document.body;
};
makeButton_fn = function(variant, icon, label, onClick) {
  const btn = document.createElement("button");
  btn.type = "button";
  btn.className = `${PANEL_CLASS}-btn ${PANEL_CLASS}-btn-${variant}`;
  btn.appendChild(createIcon(icon));
  const text = document.createElement("span");
  text.textContent = label;
  btn.appendChild(text);
  btn.addEventListener("mousedown", (e) => e.preventDefault());
  btn.addEventListener("click", (e) => {
    e.preventDefault();
    e.stopPropagation();
    onClick();
  });
  return btn;
};
makeShortcutChip_fn = function() {
  const shortcut = document.createElement("span");
  shortcut.className = `${PANEL_CLASS}-shortcut`;
  const cmd = document.createElement("span");
  cmd.textContent = __privateGet$3(this, _config).modSymbol;
  const enter = document.createElement("span");
  enter.className = `${PANEL_CLASS}-shortcut-icon`;
  setSanitizedIcon(enter, __privateGet$3(this, _config).enterKeyIcon);
  shortcut.append(cmd, enter);
  return shortcut;
};
_retry = new WeakMap();
canRetry_fn = function() {
  if (!__privateGet$3(this, _diffStartDoc)) return false;
  const editorView = this.ctx.get(core.editorViewCtx);
  return editorView.state.doc.eq(__privateGet$3(this, _diffStartDoc));
};
_rejectAll = new WeakMap();
_acceptAll = new WeakMap();

const diffActionsPanelKey = new state.PluginKey("CREPE_AI_DIFF_ACTIONS_PANEL");
const DEFAULT_DIFF_ACTIONS_RETRY_LABEL = "Retry";
const DEFAULT_DIFF_ACTIONS_REJECT_ALL_LABEL = "Reject all";
const DEFAULT_DIFF_ACTIONS_ACCEPT_ALL_LABEL = "Accept all";
function detectModSymbol() {
  var _a, _b, _c, _d;
  if (typeof navigator === "undefined") return "\u2318";
  const ua = navigator;
  const platform = (_d = (_c = (_b = (_a = ua.userAgentData) == null ? void 0 : _a.platform) != null ? _b : ua.platform) != null ? _c : ua.userAgent) != null ? _d : "";
  return /mac|iphone|ipad|ipod/i.test(platform) ? "\u2318" : "Ctrl";
}
const DEFAULT_DIFF_ACTIONS_MOD_SYMBOL = detectModSymbol();
function resolveDiffActionsConfig(options) {
  var _a, _b, _c, _d, _e, _f, _g;
  const { config, enterKeyIcon: enterKeyIcon$1 } = options;
  return {
    retryLabel: (_a = config == null ? void 0 : config.retryLabel) != null ? _a : DEFAULT_DIFF_ACTIONS_RETRY_LABEL,
    rejectAllLabel: (_b = config == null ? void 0 : config.rejectAllLabel) != null ? _b : DEFAULT_DIFF_ACTIONS_REJECT_ALL_LABEL,
    acceptAllLabel: (_c = config == null ? void 0 : config.acceptAllLabel) != null ? _c : DEFAULT_DIFF_ACTIONS_ACCEPT_ALL_LABEL,
    retryIcon: (_d = config == null ? void 0 : config.retryIcon) != null ? _d : retryIcon,
    rejectIcon: (_e = config == null ? void 0 : config.rejectIcon) != null ? _e : clearIcon,
    acceptIcon: (_f = config == null ? void 0 : config.acceptIcon) != null ? _f : confirmIcon,
    enterKeyIcon: enterKeyIcon$1 != null ? enterKeyIcon$1 : enterKeyIcon,
    modSymbol: (_g = config == null ? void 0 : config.modSymbol) != null ? _g : DEFAULT_DIFF_ACTIONS_MOD_SYMBOL
  };
}
function diffActionsPanelPlugin(options = {}) {
  const resolved = resolveDiffActionsConfig(options);
  return utils.$prose((ctx) => {
    return new state.Plugin({
      key: diffActionsPanelKey,
      view(view) {
        return new DiffActionsPanelView(ctx, view, resolved);
      },
      props: {
        handleKeyDown(view, event) {
          var _a;
          if (event.key !== "Enter") return false;
          if (!(event.metaKey || event.ctrlKey)) return false;
          if (!((_a = diff.diffPluginKey.getState(view.state)) == null ? void 0 : _a.active)) return false;
          if (!ctx.get(aiSessionCtx.key).diffOwnedByAI) return false;
          event.preventDefault();
          const commands = ctx.get(core.commandsCtx);
          commands.call(diff.acceptAllDiffsCmd.key);
          return true;
        }
      }
    });
  });
}

var __typeError$2 = (msg) => {
  throw TypeError(msg);
};
var __accessCheck$2 = (obj, member, msg) => member.has(obj) || __typeError$2("Cannot " + msg);
var __privateGet$2 = (obj, member, getter) => (__accessCheck$2(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateAdd$2 = (obj, member, value) => member.has(obj) ? __typeError$2("Cannot add the same private member more than once") : member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
var __privateSet$2 = (obj, member, value, setter) => (__accessCheck$2(obj, member, "write to private field"), member.set(obj, value), value);
var _nodes, _removeById;
function createSubmenuBuilder(node) {
  const builder = {
    addItem: (id, item) => {
      node.items.set(id, item);
      return builder;
    },
    removeItem: (id) => {
      node.items.delete(id);
      return builder;
    },
    getItem: (id) => node.items.get(id),
    clear: () => {
      node.items.clear();
      return builder;
    }
  };
  return builder;
}
class AISuggestionsBuilder {
  constructor() {
    __privateAdd$2(this, _nodes, []);
    this.addItem = (id, item) => {
      __privateGet$2(this, _removeById).call(this, id);
      __privateGet$2(this, _nodes).push({ kind: "item", id, item });
      return this;
    };
    /// Add a submenu. Populate items via the optional `build` callback,
    /// or call `getSubmenu(id)` afterward. Returns `this` so calls can be
    /// chained at the parent level alongside `addItem`.
    this.addSubmenu = (id, def, build) => {
      __privateGet$2(this, _removeById).call(this, id);
      const node = { def, items: /* @__PURE__ */ new Map() };
      if (build) build(createSubmenuBuilder(node));
      __privateGet$2(this, _nodes).push({ kind: "submenu", id, node });
      return this;
    };
    this.removeItem = (id) => {
      __privateGet$2(this, _removeById).call(this, id);
      return this;
    };
    this.getItem = (id) => {
      const node = __privateGet$2(this, _nodes).find((n) => n.kind === "item" && n.id === id);
      return (node == null ? void 0 : node.kind) === "item" ? node.item : void 0;
    };
    /// Return a builder that mutates the submenu's items in place.
    /// Multiple calls return distinct builder objects backed by the same
    /// underlying node, so changes are always visible.
    this.getSubmenu = (id) => {
      const node = __privateGet$2(this, _nodes).find((n) => n.kind === "submenu" && n.id === id);
      return (node == null ? void 0 : node.kind) === "submenu" ? createSubmenuBuilder(node.node) : void 0;
    };
    this.clear = () => {
      __privateSet$2(this, _nodes, []);
      return this;
    };
    this.build = () => {
      const main = [];
      const submenus = {};
      for (const node of __privateGet$2(this, _nodes)) {
        if (node.kind === "item") {
          main.push({ kind: "item", id: node.id, item: node.item });
        } else {
          main.push({ kind: "submenu", id: node.id, def: node.node.def });
          submenus[node.id] = {
            def: node.node.def,
            items: Array.from(node.node.items.entries()).map(([id, item]) => ({
              id,
              item
            }))
          };
        }
      }
      return { main, submenus };
    };
    __privateAdd$2(this, _removeById, (id) => {
      __privateSet$2(this, _nodes, __privateGet$2(this, _nodes).filter((n) => n.id !== id));
    });
  }
}
_nodes = new WeakMap();
_removeById = new WeakMap();
function applyDefaultSuggestions(builder) {
  builder.addItem("improve", {
    icon: aiIcon,
    label: "Improve writing",
    streamingLabel: "Improving writing",
    prompt: "Improve the writing while preserving the original meaning."
  }).addItem("grammar", {
    icon: grammarCheckIcon,
    label: "Fix grammar & spelling",
    streamingLabel: "Fixing grammar & spelling",
    prompt: "Fix any grammar and spelling errors without changing the meaning."
  }).addItem("shorter", {
    icon: shorterIcon,
    label: "Make shorter",
    streamingLabel: "Making shorter",
    prompt: "Make this shorter while preserving the key information."
  }).addItem("longer", {
    icon: longerIcon,
    label: "Make longer",
    streamingLabel: "Expanding",
    prompt: "Expand this with more detail and examples."
  });
  builder.addSubmenu(
    "tone",
    {
      icon: editIcon,
      label: "Change tone\u2026",
      title: "Change tone",
      searchPlaceholder: "Search tones\u2026"
    },
    (sub) => {
      const tones = [
        ["professional", "Professional"],
        ["casual", "Casual"],
        ["confident", "Confident"],
        ["friendly", "Friendly"],
        ["direct", "Direct"],
        ["formal", "Formal"]
      ];
      for (const [id, label] of tones) {
        sub.addItem(id, {
          icon: editIcon,
          label,
          streamingLabel: "Adjusting tone",
          prompt: `Rewrite this in a ${label.toLowerCase()} tone.`
        });
      }
    }
  );
  builder.addSubmenu(
    "translate",
    {
      icon: translateIcon,
      label: "Translate\u2026",
      title: "Translate",
      searchPlaceholder: "Search languages\u2026"
    },
    (sub) => {
      const languages = [
        ["english", "English", "English"],
        ["chinese", "Chinese", "Chinese (Simplified)"],
        ["japanese", "Japanese", "Japanese"],
        ["korean", "Korean", "Korean"],
        ["spanish", "Spanish", "Spanish"],
        ["french", "French", "French"],
        ["german", "German", "German"]
      ];
      for (const [id, label, promptName] of languages) {
        sub.addItem(id, {
          icon: translateIcon,
          label,
          streamingLabel: `Translating to ${label}`,
          prompt: `Translate this to ${promptName}.`
        });
      }
    }
  );
}

function keepAlive(..._args) {
}

keepAlive(vue.h);
function renderHighlighted(label, query) {
  const q = query.trim();
  if (!q) return [label];
  const lower = label.toLowerCase();
  const lq = q.toLowerCase();
  const idx = lower.indexOf(lq);
  if (idx === -1) return [label];
  return [
    label.slice(0, idx),
    vue.h("mark", null, label.slice(idx, idx + q.length)),
    label.slice(idx + q.length)
  ];
}
const AIInstructionInput = vue.defineComponent({
  props: {
    placeholder: { type: Object, required: true },
    resetSignal: { type: Object, required: true },
    suggestions: { type: Object, required: true },
    chrome: { type: Object, required: true },
    onConfirm: { type: Function, required: true },
    onCancel: { type: Function, required: true }
  },
  setup({
    placeholder,
    resetSignal,
    suggestions,
    chrome,
    onConfirm,
    onCancel
  }) {
    const inputValue = vue.ref("");
    const view = vue.ref({ kind: "main" });
    const selectedIndex = vue.ref(0);
    const inputRef = vue.ref(null);
    const listRef = vue.ref(null);
    const listboxId = `ai-instruction-list-${Math.random().toString(36).slice(2, 9)}`;
    const optionId = (idx) => `${listboxId}-opt-${idx}`;
    vue.watch(resetSignal, () => {
      inputValue.value = "";
      view.value = { kind: "main" };
      selectedIndex.value = 0;
    });
    const allItems = vue.computed(() => {
      if (view.value.kind === "submenu") {
        const submenu = suggestions.submenus[view.value.id];
        if (!submenu) return [];
        return submenu.items.map(({ id, item }) => ({
          id,
          icon: item.icon,
          label: item.label,
          hasSubmenu: false,
          prompt: { text: item.prompt, streamingLabel: item.streamingLabel }
        }));
      }
      return suggestions.main.map((entry) => {
        if (entry.kind === "item") {
          return {
            id: entry.id,
            icon: entry.item.icon,
            label: entry.item.label,
            hasSubmenu: false,
            prompt: {
              text: entry.item.prompt,
              streamingLabel: entry.item.streamingLabel
            }
          };
        }
        return {
          id: entry.id,
          icon: entry.def.icon,
          label: entry.def.label,
          hasSubmenu: true
        };
      });
    });
    const currentSubmenuDef = vue.computed(() => {
      var _a, _b;
      if (view.value.kind !== "submenu") return null;
      return (_b = (_a = suggestions.submenus[view.value.id]) == null ? void 0 : _a.def) != null ? _b : null;
    });
    const filteredItems = vue.computed(() => {
      const q = inputValue.value.trim().toLowerCase();
      if (!q) return allItems.value;
      return allItems.value.filter(
        (item) => item.label.toLowerCase().includes(q)
      );
    });
    const showSendAsPrompt = vue.computed(() => inputValue.value.trim().length > 0);
    const totalItems = vue.computed(
      () => filteredItems.value.length + (showSendAsPrompt.value ? 1 : 0)
    );
    vue.watch([filteredItems, view, showSendAsPrompt], () => {
      selectedIndex.value = showSendAsPrompt.value ? filteredItems.value.length : 0;
    });
    const focusInput = () => {
      void vue.nextTick(() => {
        var _a;
        return (_a = inputRef.value) == null ? void 0 : _a.focus();
      });
    };
    const enterSubmenu = (id) => {
      view.value = { kind: "submenu", id };
      inputValue.value = "";
      selectedIndex.value = 0;
      focusInput();
    };
    const exitSubmenu = () => {
      view.value = { kind: "main" };
      inputValue.value = "";
      selectedIndex.value = 0;
      focusInput();
    };
    const runItem = (item) => {
      if (item.hasSubmenu) {
        enterSubmenu(item.id);
      } else if (item.prompt) {
        onConfirm(item.prompt.text, item.prompt.streamingLabel);
        inputValue.value = "";
      }
    };
    const submitRaw = () => {
      const v = inputValue.value.trim();
      if (!v) return;
      onConfirm(v);
      inputValue.value = "";
    };
    const onSelectCurrent = () => {
      const idx = selectedIndex.value;
      const items = filteredItems.value;
      if (idx < items.length) {
        runItem(items[idx]);
      } else if (showSendAsPrompt.value) {
        submitRaw();
      }
    };
    const scrollToSelected = () => {
      void vue.nextTick(() => {
        const list = listRef.value;
        if (!list) return;
        const el = list.querySelector(
          `[data-index="${selectedIndex.value}"]`
        );
        el == null ? void 0 : el.scrollIntoView({ block: "nearest" });
      });
    };
    const onKeydown = (e) => {
      e.stopPropagation();
      if (e.key === "ArrowDown") {
        e.preventDefault();
        if (totalItems.value === 0) return;
        selectedIndex.value = (selectedIndex.value + 1) % totalItems.value;
        scrollToSelected();
        return;
      }
      if (e.key === "ArrowUp") {
        e.preventDefault();
        if (totalItems.value === 0) return;
        selectedIndex.value = (selectedIndex.value - 1 + totalItems.value) % totalItems.value;
        scrollToSelected();
        return;
      }
      if (e.key === "Enter") {
        e.preventDefault();
        onSelectCurrent();
        return;
      }
      if (e.key === "Escape") {
        e.preventDefault();
        if (view.value.kind === "submenu") exitSubmenu();
        else onCancel();
        return;
      }
      if (e.key === "Backspace" && inputValue.value === "" && view.value.kind === "submenu") {
        e.preventDefault();
        exitSubmenu();
      }
    };
    const onItemPointerDown = (e) => {
      e.preventDefault();
    };
    return () => {
      const items = filteredItems.value;
      const showPrompt = showSendAsPrompt.value;
      const submenuDef = currentSubmenuDef.value;
      return /* @__PURE__ */ vue.h("div", { class: "ai-instruction" }, /* @__PURE__ */ vue.h("div", { class: "ai-instruction-input" }, /* @__PURE__ */ vue.h("span", { class: "ai-instruction-input-prefix" }, /* @__PURE__ */ vue.h(component.Icon, { icon: chrome.aiIcon })), /* @__PURE__ */ vue.h(
        "input",
        {
          ref: inputRef,
          class: "ai-instruction-input-field",
          role: "combobox",
          "aria-expanded": "true",
          "aria-autocomplete": "list",
          "aria-controls": listboxId,
          "aria-activedescendant": totalItems.value > 0 ? optionId(selectedIndex.value) : void 0,
          placeholder: submenuDef ? submenuDef.searchPlaceholder : placeholder.value,
          value: inputValue.value,
          onInput: (e) => {
            inputValue.value = e.target.value;
          },
          onKeydown
        }
      ), /* @__PURE__ */ vue.h(
        "button",
        {
          type: "button",
          class: "ai-instruction-submit",
          "aria-label": chrome.submitButtonLabel,
          disabled: !showPrompt,
          onMousedown: onItemPointerDown,
          onClick: submitRaw
        },
        /* @__PURE__ */ vue.h(component.Icon, { icon: chrome.sendIcon })
      )), /* @__PURE__ */ vue.h(
        "div",
        {
          class: "ai-instruction-list",
          ref: listRef,
          id: listboxId,
          role: "listbox",
          "aria-label": chrome.listboxLabel
        },
        submenuDef && /* @__PURE__ */ vue.h(
          "button",
          {
            type: "button",
            class: "ai-instruction-back",
            onMousedown: onItemPointerDown,
            onClick: exitSubmenu
          },
          /* @__PURE__ */ vue.h("span", { class: "ai-instruction-back-icon", "aria-hidden": "true" }, /* @__PURE__ */ vue.h(component.Icon, { icon: chrome.chevronLeftIcon })),
          /* @__PURE__ */ vue.h("span", null, submenuDef.title)
        ),
        items.length > 0 && /* @__PURE__ */ vue.h("div", { class: "ai-instruction-section" }, /* @__PURE__ */ vue.h("div", { class: "ai-instruction-section-header" }, chrome.suggestionsHeaderLabel), items.map((item, idx) => /* @__PURE__ */ vue.h(
          "div",
          {
            key: item.id,
            id: optionId(idx),
            "data-index": idx,
            role: "option",
            "aria-selected": idx === selectedIndex.value,
            class: [
              "ai-instruction-item",
              idx === selectedIndex.value ? "active" : ""
            ],
            onMousedown: onItemPointerDown,
            onClick: () => runItem(item),
            onPointerenter: () => {
              selectedIndex.value = idx;
            }
          },
          /* @__PURE__ */ vue.h("span", { class: "ai-instruction-item-icon" }, /* @__PURE__ */ vue.h(component.Icon, { icon: item.icon })),
          /* @__PURE__ */ vue.h("span", { class: "ai-instruction-item-label" }, renderHighlighted(item.label, inputValue.value)),
          item.hasSubmenu && /* @__PURE__ */ vue.h("span", { class: "ai-instruction-item-arrow" }, /* @__PURE__ */ vue.h(component.Icon, { icon: chrome.chevronRightIcon }))
        ))),
        showPrompt && /* @__PURE__ */ vue.h("div", { class: "ai-instruction-section" }, /* @__PURE__ */ vue.h("div", { class: "ai-instruction-section-header" }, chrome.sendAsPromptHeaderLabel), /* @__PURE__ */ vue.h(
          "div",
          {
            id: optionId(items.length),
            "data-index": items.length,
            role: "option",
            "aria-selected": selectedIndex.value === items.length,
            class: [
              "ai-instruction-item",
              "ai-instruction-item-prompt",
              selectedIndex.value === items.length ? "active" : ""
            ],
            onMousedown: onItemPointerDown,
            onClick: submitRaw,
            onPointerenter: () => {
              selectedIndex.value = items.length;
            }
          },
          /* @__PURE__ */ vue.h("span", { class: "ai-instruction-item-icon" }, /* @__PURE__ */ vue.h(component.Icon, { icon: chrome.sendPromptIcon })),
          /* @__PURE__ */ vue.h("span", { class: "ai-instruction-item-label" }, chrome.sendAsPromptLabel, " ", /* @__PURE__ */ vue.h("span", { class: "ai-instruction-item-quote" }, '"', inputValue.value, '"')),
          /* @__PURE__ */ vue.h("span", { class: "ai-instruction-item-shortcut" }, /* @__PURE__ */ vue.h(component.Icon, { icon: chrome.enterKeyIcon }))
        ))
      ));
    };
  }
});

var __typeError$1 = (msg) => {
  throw TypeError(msg);
};
var __accessCheck$1 = (obj, member, msg) => member.has(obj) || __typeError$1("Cannot " + msg);
var __privateGet$1 = (obj, member, getter) => (__accessCheck$1(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateAdd$1 = (obj, member, value) => member.has(obj) ? __typeError$1("Cannot add the same private member more than once") : member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
var __privateSet$1 = (obj, member, value, setter) => (__accessCheck$1(obj, member, "write to private field"), member.set(obj, value), value);
var _content, _provider, _app, _placeholder, _resetSignal, _from, _to, _wantsShow, _onConfirm, _onCancel;
class AIInstructionTooltipView {
  constructor(ctx, view, config) {
    this.ctx = ctx;
    __privateAdd$1(this, _content);
    __privateAdd$1(this, _provider);
    __privateAdd$1(this, _app);
    __privateAdd$1(this, _placeholder);
    __privateAdd$1(this, _resetSignal);
    __privateAdd$1(this, _from, -1);
    __privateAdd$1(this, _to, -1);
    /// Source of truth for "should the palette currently be showing".
    /// `TooltipProvider.shouldShow` reads this so that forwarding `update()`
    /// to the provider on every editor transition (needed to keep the
    /// floating position in sync with layout changes) doesn't dismiss the
    /// palette behind our back.
    __privateAdd$1(this, _wantsShow, false);
    __privateAdd$1(this, _onConfirm, (instruction, label) => {
      if (!instruction.trim()) return;
      const commands = this.ctx.get(core.commandsCtx);
      const accepted = commands.call(runAICmd.key, { instruction, label });
      if (!accepted) return;
      __privateSet$1(this, _wantsShow, false);
      __privateGet$1(this, _provider).hide();
    });
    __privateAdd$1(this, _onCancel, () => {
      __privateSet$1(this, _wantsShow, false);
      __privateGet$1(this, _provider).hide();
    });
    this.show = (from, to) => {
      __privateSet$1(this, _from, from);
      __privateSet$1(this, _to, to);
      __privateGet$1(this, _resetSignal).value++;
      __privateSet$1(this, _wantsShow, true);
      const view = this.ctx.get(core.editorViewCtx);
      __privateGet$1(this, _provider).show(
        { getBoundingClientRect: () => prose.posToDOMRect(view, from, to) },
        view
      );
      requestAnimationFrame(() => {
        var _a;
        (_a = __privateGet$1(this, _content).querySelector("input")) == null ? void 0 : _a.focus();
      });
    };
    this.update = (view, prevState) => {
      const { selection } = view.state;
      const isTextSelection = selection instanceof state.TextSelection;
      const movedRange = selection.from !== __privateGet$1(this, _from) || selection.to !== __privateGet$1(this, _to);
      if (!isTextSelection || movedRange) {
        __privateSet$1(this, _wantsShow, false);
        __privateGet$1(this, _provider).hide();
        return;
      }
      __privateGet$1(this, _provider).update(view, prevState);
    };
    this.destroy = () => {
      __privateGet$1(this, _app).unmount();
      __privateGet$1(this, _provider).destroy();
      __privateGet$1(this, _content).remove();
    };
    __privateSet$1(this, _placeholder, vue.ref(config.placeholder));
    __privateSet$1(this, _resetSignal, vue.ref(0));
    const content = document.createElement("div");
    content.className = "milkdown-ai-instruction";
    const app = vue.createApp(AIInstructionInput, {
      placeholder: __privateGet$1(this, _placeholder),
      resetSignal: __privateGet$1(this, _resetSignal),
      suggestions: config.suggestions,
      chrome: config.chrome,
      onConfirm: __privateGet$1(this, _onConfirm),
      onCancel: __privateGet$1(this, _onCancel)
    });
    app.mount(content);
    __privateSet$1(this, _app, app);
    __privateSet$1(this, _content, content);
    __privateSet$1(this, _provider, new tooltip.TooltipProvider({
      content,
      debounce: 0,
      offset: 10,
      shouldShow: () => __privateGet$1(this, _wantsShow),
      floatingUIOptions: {
        placement: "bottom"
      }
    }));
    __privateGet$1(this, _provider).onHide = () => {
      __privateSet$1(this, _wantsShow, false);
      requestAnimationFrame(() => {
        try {
          const root = __privateGet$1(this, _content).getRootNode();
          const active = root.activeElement;
          if (!active || !__privateGet$1(this, _content).contains(active)) return;
          const v = this.ctx.get(core.editorViewCtx);
          v.dom.focus({ preventScroll: true });
        } catch (e) {
        }
      });
    };
    __privateGet$1(this, _provider).update(view);
  }
}
_content = new WeakMap();
_provider = new WeakMap();
_app = new WeakMap();
_placeholder = new WeakMap();
_resetSignal = new WeakMap();
_from = new WeakMap();
_to = new WeakMap();
_wantsShow = new WeakMap();
_onConfirm = new WeakMap();
_onCancel = new WeakMap();

const defaultAPI = {
  show: () => {
  }
};
const aiInstructionTooltipAPI = utils.$ctx(
  { ...defaultAPI },
  "aiInstructionTooltipAPI"
);
const aiInstructionTooltip = tooltip.tooltipFactory("CREPE_AI_INSTRUCTION");
const DEFAULT_SUGGESTIONS_HEADER_LABEL = "SUGGESTIONS";
const DEFAULT_SEND_AS_PROMPT_HEADER_LABEL = "SEND AS PROMPT";
const DEFAULT_SEND_AS_PROMPT_LABEL = "Ask AI:";
const DEFAULT_SUBMIT_BUTTON_LABEL = "Send prompt";
const DEFAULT_LISTBOX_LABEL = "AI suggestions";
const DEFAULT_INSTRUCTION_PLACEHOLDER = "Tell AI what to do with the selection\u2026";
function resolveChrome(config) {
  var _a, _b, _c, _d, _e, _f, _g, _h, _i, _j, _k;
  return {
    aiIcon: (_a = config == null ? void 0 : config.aiIcon) != null ? _a : aiIcon,
    sendIcon: (_b = config == null ? void 0 : config.sendIcon) != null ? _b : sendIcon,
    sendPromptIcon: (_c = config == null ? void 0 : config.sendPromptIcon) != null ? _c : sendPromptIcon,
    enterKeyIcon: (_d = config == null ? void 0 : config.enterKeyIcon) != null ? _d : enterKeyIcon,
    chevronLeftIcon: (_e = config == null ? void 0 : config.chevronLeftIcon) != null ? _e : chevronLeftIcon,
    chevronRightIcon: (_f = config == null ? void 0 : config.chevronRightIcon) != null ? _f : chevronRightIcon,
    suggestionsHeaderLabel: (_g = config == null ? void 0 : config.suggestionsHeaderLabel) != null ? _g : DEFAULT_SUGGESTIONS_HEADER_LABEL,
    sendAsPromptHeaderLabel: (_h = config == null ? void 0 : config.sendAsPromptHeaderLabel) != null ? _h : DEFAULT_SEND_AS_PROMPT_HEADER_LABEL,
    sendAsPromptLabel: (_i = config == null ? void 0 : config.sendAsPromptLabel) != null ? _i : DEFAULT_SEND_AS_PROMPT_LABEL,
    submitButtonLabel: (_j = config == null ? void 0 : config.submitButtonLabel) != null ? _j : DEFAULT_SUBMIT_BUTTON_LABEL,
    listboxLabel: (_k = config == null ? void 0 : config.listboxLabel) != null ? _k : DEFAULT_LISTBOX_LABEL
  };
}
function resolveViewConfig(config) {
  var _a, _b;
  const builder = new AISuggestionsBuilder();
  applyDefaultSuggestions(builder);
  (_a = config == null ? void 0 : config.buildAISuggestions) == null ? void 0 : _a.call(config, builder);
  return {
    placeholder: (_b = config == null ? void 0 : config.instructionPlaceholder) != null ? _b : DEFAULT_INSTRUCTION_PLACEHOLDER,
    chrome: resolveChrome(config),
    suggestions: builder.build()
  };
}
function configureAIInstructionTooltip(config) {
  return (ctx) => {
    const viewConfig = resolveViewConfig(config);
    let tooltipView = null;
    ctx.update(aiInstructionTooltipAPI.key, (api) => ({
      ...api,
      show: (from, to) => {
        tooltipView == null ? void 0 : tooltipView.show(from, to);
      }
    }));
    ctx.set(aiInstructionTooltip.key, {
      view: (view) => {
        tooltipView = new AIInstructionTooltipView(ctx, view, viewConfig);
        return tooltipView;
      }
    });
  };
}

var __typeError = (msg) => {
  throw TypeError(msg);
};
var __accessCheck = (obj, member, msg) => member.has(obj) || __typeError("Cannot " + msg);
var __privateGet = (obj, member, getter) => (__accessCheck(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateAdd = (obj, member, value) => member.has(obj) ? __typeError("Cannot add the same private member more than once") : member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
var __privateSet = (obj, member, value, setter) => (__accessCheck(obj, member, "write to private field"), member.set(obj, value), value);
var _spinner, _label, _fallbackLabel, _start, _lastLabelText, _rafId, _tick;
const CLASS_PREFIX = "milkdown-ai-streaming";
const SPINNER_PERIOD_MS = 800;
const DEFAULT_STREAMING_FALLBACK_LABEL = "Generating";
const DEFAULT_STREAMING_CANCEL_HINT = "Esc to cancel";
const indicatorKey = new state.PluginKey(
  "CREPE_AI_STREAMING_INDICATOR"
);
class IndicatorWidget {
  constructor(ctx, fallbackLabel, cancelHint) {
    __privateAdd(this, _spinner);
    __privateAdd(this, _label);
    __privateAdd(this, _fallbackLabel);
    __privateAdd(this, _start, performance.now());
    __privateAdd(this, _lastLabelText, "");
    __privateAdd(this, _rafId, 0);
    __privateAdd(this, _tick, () => {
      const elapsed = performance.now() - __privateGet(this, _start);
      const angle = elapsed / SPINNER_PERIOD_MS * 360;
      __privateGet(this, _spinner).style.transform = `rotate(${angle}deg)`;
      __privateSet(this, _rafId, requestAnimationFrame(__privateGet(this, _tick)));
    });
    __privateSet(this, _fallbackLabel, fallbackLabel);
    const dom = document.createElement("span");
    dom.className = `${CLASS_PREFIX}-indicator`;
    dom.contentEditable = "false";
    dom.setAttribute("role", "status");
    dom.setAttribute("aria-live", "polite");
    const spinner = document.createElement("span");
    spinner.className = `${CLASS_PREFIX}-spinner`;
    spinner.setAttribute("aria-hidden", "true");
    dom.appendChild(spinner);
    const label = document.createElement("span");
    label.className = `${CLASS_PREFIX}-label`;
    dom.appendChild(label);
    const escHint = document.createElement("span");
    escHint.className = `${CLASS_PREFIX}-esc`;
    escHint.textContent = cancelHint;
    dom.appendChild(escHint);
    this.dom = dom;
    __privateSet(this, _spinner, spinner);
    __privateSet(this, _label, label);
    this.setLabel(ctx);
    __privateGet(this, _tick).call(this);
  }
  setLabel(ctx) {
    const session = ctx.get(aiSessionCtx.key);
    const text = `${session.label || __privateGet(this, _fallbackLabel)}\u2026`;
    if (text === __privateGet(this, _lastLabelText)) return;
    __privateSet(this, _lastLabelText, text);
    __privateGet(this, _label).textContent = text;
  }
  destroy() {
    if (__privateGet(this, _rafId)) cancelAnimationFrame(__privateGet(this, _rafId));
    __privateSet(this, _rafId, 0);
  }
}
_spinner = new WeakMap();
_label = new WeakMap();
_fallbackLabel = new WeakMap();
_start = new WeakMap();
_lastLabelText = new WeakMap();
_rafId = new WeakMap();
_tick = new WeakMap();
function streamingIndicatorPlugin(options = {}) {
  var _a, _b;
  const { config } = options;
  const fallbackLabel = (_a = config == null ? void 0 : config.fallbackLabel) != null ? _a : DEFAULT_STREAMING_FALLBACK_LABEL;
  const cancelHint = (_b = config == null ? void 0 : config.cancelHint) != null ? _b : DEFAULT_STREAMING_CANCEL_HINT;
  return utils.$prose((ctx) => {
    let widget = null;
    function ensureWidget() {
      if (!widget) widget = new IndicatorWidget(ctx, fallbackLabel, cancelHint);
      else widget.setLabel(ctx);
      return widget;
    }
    function dropWidget() {
      if (widget) {
        widget.destroy();
        widget = null;
      }
    }
    function buildSet(doc, insertEndPos) {
      const pos = Math.min(insertEndPos, doc.content.size);
      const decoration = view.Decoration.widget(pos, ensureWidget().dom, {
        side: 1,
        key: "ai-streaming-indicator"
      });
      return view.DecorationSet.create(doc, [decoration]);
    }
    return new state.Plugin({
      key: indicatorKey,
      state: {
        init: () => view.DecorationSet.empty,
        apply(tr, decorations, _oldState, newState) {
          const streaming$1 = streaming.streamingPluginKey.getState(newState);
          if (!(streaming$1 == null ? void 0 : streaming$1.active) || streaming$1.insertEndPos == null) {
            dropWidget();
            return view.DecorationSet.empty;
          }
          if (tr.getMeta(streaming.streamingPluginKey) || tr.docChanged) {
            return buildSet(newState.doc, streaming$1.insertEndPos);
          }
          return decorations.map(tr.mapping, tr.doc);
        }
      },
      view() {
        return {
          destroy() {
            dropWidget();
          }
        };
      },
      props: {
        decorations(state) {
          var _a2;
          return (_a2 = indicatorKey.getState(state)) != null ? _a2 : view.DecorationSet.empty;
        },
        handleKeyDown(view, event) {
          var _a2;
          if (event.key !== "Escape") return false;
          const commands = ctx.get(core.commandsCtx);
          if (ctx.get(aiSessionCtx.key).abortController) {
            event.preventDefault();
            commands.call(abortAICmd.key, { keep: true });
            return true;
          }
          if ((_a2 = streaming.streamingPluginKey.getState(view.state)) == null ? void 0 : _a2.active) {
            event.preventDefault();
            commands.call(streaming.abortStreamingCmd.key, { keep: true });
            return true;
          }
          return false;
        }
      }
    });
  });
}

const CREPE_CUSTOM_BLOCK_TYPES = ["table", "image-block", "code_block"];
const CREPE_IGNORE_ATTRS = { heading: ["id"] };
const ai = (editor, config) => {
  var _a, _b;
  const diffCfg = (_a = config == null ? void 0 : config.diff) != null ? _a : {};
  const streamingCfg = (_b = config == null ? void 0 : config.streaming) != null ? _b : {};
  editor.config(crepeFeatureConfig(CrepeFeature.AI)).config((ctx) => {
    ctx.update(diff.diffConfig.key, (prev) => {
      var _a2;
      return {
        ...prev,
        ignoreAttrs: (_a2 = diffCfg.ignoreAttrs) != null ? _a2 : CREPE_IGNORE_ATTRS
      };
    });
    const { ignoreAttrs: _, ...componentConfig } = diffCfg;
    ctx.update(diff$1.diffComponentConfig.key, (prev) => {
      var _a2;
      return {
        ...prev,
        customBlockTypes: (_a2 = componentConfig.customBlockTypes) != null ? _a2 : CREPE_CUSTOM_BLOCK_TYPES,
        ...componentConfig
      };
    });
  }).use(diff.diff).use(diff$1.diffComponent).config((ctx) => {
    ctx.update(streaming.streamingConfig.key, (prev) => {
      var _a2;
      return {
        ...prev,
        ...streamingCfg,
        ignoreAttrs: (_a2 = streamingCfg.ignoreAttrs) != null ? _a2 : CREPE_IGNORE_ATTRS,
        // Wire diffReviewOnEnd into the streaming plugin so manual
        // endStreamingCmd calls (outside runAICmd) also respect it.
        // Only override if the user explicitly set it — otherwise
        // keep the streaming plugin's own default so tests and
        // manual-streaming use cases aren't surprised.
        ...(config == null ? void 0 : config.diffReviewOnEnd) !== void 0 ? { diffReviewOnEnd: config.diffReviewOnEnd } : {}
      };
    });
  }).use(streaming.streaming).config((ctx) => {
    ctx.update(aiProviderConfig.key, (prev) => {
      var _a2;
      return {
        ...prev,
        ...(config == null ? void 0 : config.provider) !== void 0 ? { provider: config.provider } : {},
        ...(config == null ? void 0 : config.buildContext) !== void 0 ? { buildContext: config.buildContext } : {},
        diffReviewOnEnd: (_a2 = config == null ? void 0 : config.diffReviewOnEnd) != null ? _a2 : prev.diffReviewOnEnd,
        ...(config == null ? void 0 : config.onError) !== void 0 ? { onError: config.onError } : {},
        ...(config == null ? void 0 : config.aiIcon) !== void 0 ? { aiIcon: config.aiIcon } : {}
      };
    });
  }).use(aiProviderConfig).use(aiSessionCtx).use(runAICmd).use(abortAICmd).config(configureAIInstructionTooltip(config)).use(aiInstructionTooltipAPI).use(aiInstructionTooltip).use(streamingIndicatorPlugin({ config: config == null ? void 0 : config.streamingIndicator })).use(
    diffActionsPanelPlugin({
      config: config == null ? void 0 : config.diffActions,
      enterKeyIcon: config == null ? void 0 : config.enterKeyIcon
    })
  );
};

exports.abortAICmd = abortAICmd;
exports.ai = ai;
exports.defaultBuildContext = defaultBuildContext;
exports.runAICmd = runAICmd;
