// vite-plugin-ruby/dist/index.mjs
import path from "node:path";
import { fileURLToPath } from "node:url";
import path$1, { basename, join, posix, relative, resolve } from "path";
import { existsSync, promises, readFileSync } from "fs";
import { createDebug } from "obug";
import { globSync } from "tinyglobby";
//#region ../node_modules/.pnpm/tsdown@0.21.4_typescript@5.9.3/node_modules/tsdown/esm-shims.js
const getFilename = () => fileURLToPath(import.meta.url);
const getDirname = () => path.dirname(getFilename());
const __dirname = /* @__PURE__ */ getDirname();
//#endregion
//#region src/constants.ts
const APP_ENV = process.env.RAILS_ENV || process.env.RACK_ENV || process.env.APP_ENV;
const ENV_PREFIX = "VITE_RUBY";
const KNOWN_ENTRYPOINT_TYPES = [
	"html",
	"jsx?",
	"tsx?",
	...[
		"css",
		"less",
		"sass",
		"scss",
		"styl",
		"stylus",
		"pcss",
		"postcss"
	]
];
const ENTRYPOINT_TYPES_REGEX = new RegExp(`\\.(${KNOWN_ENTRYPOINT_TYPES.join("|")})(\\?.*)?$`);
//#endregion
//#region src/utils.ts
function slash(path) {
	return path.replace(/\\/g, "/");
}
function isObject(value) {
	return Object.prototype.toString.call(value) === "[object Object]";
}
function screamCase(key) {
	return key.replace(/([a-z])([A-Z])/g, "$1_$2").toUpperCase();
}
function configOptionFromEnv(optionName) {
	return process.env[`${ENV_PREFIX}_${screamCase(optionName)}`];
}
function booleanOption(value) {
	if (value === "true") return true;
	if (value === "false") return false;
	return value;
}
function loadJsonConfig(filepath) {
	return JSON.parse(readFileSync(filepath, {
		encoding: "utf8",
		flag: "r"
	}));
}
function cleanConfig(object) {
	Object.keys(object).forEach((key) => {
		const value = object[key];
		if (value === void 0 || value === null) delete object[key];
		else if (isObject(value)) cleanConfig(value);
	});
	return object;
}
//#endregion
//#region src/config.ts
const defaultConfig = loadJsonConfig(resolve(__dirname, "../default.vite.json"));
function filterEntrypointsForRollup(entrypoints) {
	return entrypoints.filter(([_name, filename]) => ENTRYPOINT_TYPES_REGEX.test(filename));
}
function filterEntrypointAssets(entrypoints) {
	return entrypoints.filter(([_name, filename]) => !ENTRYPOINT_TYPES_REGEX.test(filename));
}
function resolveEntrypointFiles(projectRoot, sourceCodeDir, config) {
	const entrypointFiles = globSync(resolveGlobs(projectRoot, sourceCodeDir, config.ssrBuild ? [config.ssrEntrypoint] : [`~/${config.entrypointsDir}/**/*`, ...config.additionalEntrypoints]), {
		absolute: true,
		expandDirectories: false
	});
	if (config.ssrBuild) {
		if (entrypointFiles.length === 0) throw new Error(`No SSR entrypoint available, please create \`${config.ssrEntrypoint}\` to do an SSR build.`);
		else if (entrypointFiles.length > 1) throw new Error(`Expected a single SSR entrypoint, found: ${entrypointFiles}`);
		return entrypointFiles.map((file) => ["ssr", file]);
	}
	return entrypointFiles.map((filename) => {
		let name = relative(sourceCodeDir, filename);
		if (name.startsWith("..")) name = relative(projectRoot, filename);
		return [name, filename];
	});
}
function resolveGlobs(projectRoot, sourceCodeDir, patterns) {
	return patterns.map((pattern) => slash(resolve(projectRoot, pattern.replace(/^~\//, `${sourceCodeDir}/`))));
}
function configFromEnv() {
	const envConfig = {};
	Object.keys(defaultConfig).forEach((optionName) => {
		const envValue = configOptionFromEnv(optionName);
		if (envValue !== void 0) envConfig[optionName] = envValue;
	});
	return envConfig;
}
function loadConfiguration(viteMode, projectRoot, userConfig) {
	const envConfig = configFromEnv();
	const mode = envConfig.mode || APP_ENV || viteMode;
	const multiEnvConfig = loadJsonConfig(join(projectRoot, envConfig.configPath || defaultConfig.configPath));
	const fileConfig = {
		...multiEnvConfig["all"],
		...multiEnvConfig[mode]
	};
	return coerceConfigurationValues({
		...defaultConfig,
		...fileConfig,
		...envConfig,
		mode
	}, projectRoot, userConfig);
}
function coerceConfigurationValues(config, projectRoot, userConfig) {
	const port = config.port = parseInt(config.port);
	const https = config.https = userConfig.server?.https || booleanOption(config.https);
	const server = {
		fs: {
			allow: [projectRoot],
			strict: userConfig.server?.fs?.strict ?? true
		},
		host: config.host,
		https,
		port,
		strictPort: true
	};
	if (booleanOption(config.skipProxy)) server.origin = userConfig.server?.origin || `${https ? "https" : "http"}://${config.host}:${config.port}`;
	const hmr = userConfig.server?.hmr ?? {};
	if (typeof hmr === "object" && !hmr.hasOwnProperty("clientPort")) {
		hmr.clientPort ||= port;
		server.hmr = hmr;
	}
	const root = join(projectRoot, config.sourceCodeDir);
	const ssrEntrypoint = userConfig.build?.ssr;
	config.ssrBuild = Boolean(ssrEntrypoint);
	if (typeof ssrEntrypoint === "string") config.ssrEntrypoint = ssrEntrypoint;
	const outDir = relative(root, config.ssrBuild ? config.ssrOutputDir : join(config.publicDir, config.publicOutputDir));
	const base = resolveViteBase(config);
	const entrypoints = resolveEntrypointFiles(projectRoot, root, config);
	return {
		...config,
		server,
		root,
		outDir,
		base,
		entrypoints
	};
}
function resolveViteBase({ assetHost, base, publicOutputDir }) {
	if (assetHost && !assetHost.startsWith("http")) assetHost = `//${assetHost}`;
	return [ensureTrailingSlash(assetHost || base || "/"), publicOutputDir ? ensureTrailingSlash(slash(publicOutputDir)) : ""].join("");
}
function ensureTrailingSlash(path) {
	return path.endsWith("/") ? path : `${path}/`;
}
//#endregion
//#region src/manifest.ts
const debug$1 = createDebug("vite-plugin-ruby:assets-manifest");
function assetsManifestPlugin() {
	let config;
	let viteRubyConfig;
	async function fingerprintRemainingAssets(ctx, bundle, manifest) {
		const remainingAssets = filterEntrypointAssets(viteRubyConfig.entrypoints);
		for (const [filename, absoluteFilename] of remainingAssets) {
			const content = await promises.readFile(absoluteFilename);
			const ref = ctx.emitFile({
				name: path$1.basename(filename),
				type: "asset",
				source: content
			});
			const hashedFilename = ctx.getFileName(ref);
			manifest.set(path$1.relative(config.root, absoluteFilename), {
				file: hashedFilename,
				src: filename
			});
		}
	}
	return {
		name: "vite-plugin-ruby:assets-manifest",
		apply: "build",
		enforce: "post",
		configResolved(resolvedConfig) {
			config = resolvedConfig;
			viteRubyConfig = config.viteRuby;
		},
		async generateBundle(_options, bundle) {
			if (!config.build.manifest) return;
			const fileName = `${typeof config.build.manifest === "string" ? path$1.dirname(config.build.manifest) : ".vite"}/manifest-assets.json`;
			const manifest = /* @__PURE__ */ new Map();
			await fingerprintRemainingAssets(this, bundle, manifest);
			debug$1({
				manifest,
				fileName
			});
			this.emitFile({
				fileName,
				type: "asset",
				source: JSON.stringify(Object.fromEntries(manifest), null, 2)
			});
		}
	};
}
//#endregion
//#region src/index.ts
const projectRoot = configOptionFromEnv("root") || process.cwd();
let watchAdditionalPaths = [];
function ViteRubyPlugin() {
	return [{
		name: "vite-plugin-ruby",
		config,
		configureServer
	}, assetsManifestPlugin()];
}
const debug = createDebug("vite-plugin-ruby:config");
function config(userConfig, env) {
	const config = loadConfiguration(env.mode, projectRoot, userConfig);
	const { assetsDir, base, outDir, server, root, entrypoints, ssrBuild } = config;
	const isLocal = config.mode === "development" || config.mode === "test";
	const rollupOptions = userConfig.build?.rollupOptions;
	let rollupInput = rollupOptions?.input;
	if (typeof rollupInput === "string") rollupInput = { [rollupInput]: rollupInput };
	const isUsingRolldown = this && this.meta && this.meta.rolldownVersion;
	const optionsKey = isUsingRolldown ? "rolldownOptions" : "rollupOptions";
	const build = {
		emptyOutDir: userConfig.build?.emptyOutDir ?? (ssrBuild || isLocal),
		sourcemap: !isLocal,
		...userConfig.build,
		assetsDir,
		manifest: !ssrBuild,
		outDir,
		...isUsingRolldown && { rollupOptions: void 0 },
		[optionsKey]: {
			...rollupOptions,
			input: {
				...rollupInput,
				...Object.fromEntries(filterEntrypointsForRollup(entrypoints))
			},
			output: {
				...outputOptions(assetsDir, ssrBuild),
				...rollupOptions?.output
			}
		}
	};
	const envDir = userConfig.envDir || projectRoot;
	debug({
		base,
		build,
		envDir,
		root,
		server,
		entrypoints: Object.fromEntries(entrypoints)
	});
	watchAdditionalPaths = resolveGlobs(projectRoot, root, config.watchAdditionalPaths || []);
	return cleanConfig({
		resolve: { alias: {
			"~/": `${root}/`,
			"@/": `${root}/`
		} },
		base,
		envDir,
		root,
		server,
		build,
		viteRuby: config
	});
}
function configureServer(server) {
	server.watcher.add(watchAdditionalPaths);
	return () => server.middlewares.use((req, res, next) => {
		if (req.url === "/index.html" && !existsSync(resolve(server.config.root, "index.html"))) {
			res.statusCode = 404;
			const file = readFileSync(resolve(__dirname, "dev-server-index.html"), "utf-8");
			res.end(file);
		}
		next();
	});
}
function outputOptions(assetsDir, ssrBuild) {
	const outputFileName = (ext) => (asset) => {
		const shortName = basename(asset.names?.[0] ?? asset.name ?? "[name]").split(".")[0];
		return posix.join(assetsDir, `${shortName}-[hash].${ext}`);
	};
	return {
		assetFileNames: ssrBuild ? void 0 : outputFileName("[ext]"),
		entryFileNames: ssrBuild ? void 0 : outputFileName("js")
	};
}
//#endregion
export { ViteRubyPlugin as default, projectRoot };
