// @milkdown/crepe/lib/esm/llm-providers/anthropic/index.js
const DEFAULT_SYSTEM_PROMPT = `You are a writing assistant embedded in a markdown editor.

Rules:
- Output markdown only. Never wrap your output in code fences (e.g. \`\`\`markdown ... \`\`\`).
- Never include preambles, explanations, or sign-offs \u2014 output only the edited or generated content itself.
- Preserve the original markdown structure (headings, lists, links, code blocks) unless the instruction explicitly asks to change it.
- If a <selection> is provided, return only the replacement for that selection \u2014 do not repeat surrounding document context.
- If no <selection> is provided, return content to insert at the cursor that flows with the surrounding document.`;
function buildDefaultUserMessage(context) {
  const parts = [`<document>
${context.document}
</document>`];
  if (context.selection) {
    parts.push(`<selection>
${context.selection}
</selection>`);
  }
  parts.push(`<instruction>
${context.instruction}
</instruction>`);
  return parts.join("\n\n");
}
function isBrowserLike() {
  const g = globalThis;
  if (g.document !== void 0) return true;
  if (g.WorkerGlobalScope !== void 0) return true;
  return false;
}
function assertBrowserSafe(config, providerName) {
  if (!config.apiKey) return;
  if (!isBrowserLike()) return;
  if (config.dangerouslyAllowBrowser) return;
  throw new Error(
    `[${providerName}] Refusing to send your API key from a browser. Direct browser \u2192 provider calls expose the key to every visitor. Either route through your backend (set \`baseURL\` + \`headers\` and omit \`apiKey\`), or set \`dangerouslyAllowBrowser: true\` to acknowledge the risk (recommended only for desktop apps or BYOK setups).`
  );
}
function resolveSystemPrompt(systemPrompt) {
  if (systemPrompt === null) return null;
  return systemPrompt != null ? systemPrompt : DEFAULT_SYSTEM_PROMPT;
}
function stripTrailingSlashes(url) {
  let end = url.length;
  while (end > 0 && url.charCodeAt(end - 1) === 47) end--;
  return end === url.length ? url : url.slice(0, end);
}
async function* parseSSE(response, signal) {
  if (!response.body) return;
  const reader = response.body.getReader();
  const decoder = new TextDecoder("utf-8");
  let buffer = "";
  try {
    while (true) {
      if (signal.aborted) return;
      const { done, value } = await reader.read();
      if (signal.aborted) return;
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      let nl;
      while ((nl = buffer.indexOf("\n")) >= 0) {
        const raw = buffer.slice(0, nl);
        buffer = buffer.slice(nl + 1);
        const line = raw.endsWith("\r") ? raw.slice(0, -1) : raw;
        if (line.startsWith("data: ")) {
          yield line.slice(6);
        } else if (line.startsWith("data:")) {
          yield line.slice(5);
        }
      }
    }
    const tail = buffer.endsWith("\r") ? buffer.slice(0, -1) : buffer;
    if (tail.startsWith("data: ")) yield tail.slice(6);
    else if (tail.startsWith("data:")) yield tail.slice(5);
  } finally {
    reader.releaseLock();
  }
}
async function readErrorBody(response, providerName) {
  let body = "";
  try {
    body = await response.text();
  } catch (e) {
  }
  return new Error(
    `[${providerName}] Request failed with status ${response.status}` + (body ? `: ${body.slice(0, 500)}` : "")
  );
}

const PROVIDER_NAME = "milkdown/providers/anthropic";
const DEFAULT_BASE_URL = "https://api.anthropic.com";
const DEFAULT_ANTHROPIC_VERSION = "2023-06-01";
const DEFAULT_MAX_TOKENS = 4096;
function createAnthropicProvider(config) {
  assertBrowserSafe(config, PROVIDER_NAME);
  return async function* anthropicProvider(context, signal) {
    var _a, _b, _c, _d;
    const systemPrompt = resolveSystemPrompt(config.systemPrompt);
    const userMessage = buildDefaultUserMessage(context);
    const built = config.buildMessages ? config.buildMessages(context, { systemPrompt, userMessage }) : {
      system: systemPrompt != null ? systemPrompt : void 0,
      messages: [
        { role: "user", content: userMessage }
      ]
    };
    const baseURL = stripTrailingSlashes((_a = config.baseURL) != null ? _a : DEFAULT_BASE_URL);
    const url = `${baseURL}/v1/messages`;
    const headers = {
      "Content-Type": "application/json",
      "anthropic-version": (_b = config.anthropicVersion) != null ? _b : DEFAULT_ANTHROPIC_VERSION,
      ...config.apiKey ? { "x-api-key": config.apiKey } : {},
      // Required for direct browser → api.anthropic.com calls; harmless
      // when proxying through a backend.
      ...config.dangerouslyAllowBrowser ? { "anthropic-dangerous-direct-browser-access": "true" } : {},
      ...config.headers
    };
    const body = JSON.stringify({
      model: config.model,
      max_tokens: (_c = config.maxTokens) != null ? _c : DEFAULT_MAX_TOKENS,
      stream: true,
      // Caller-supplied empty string is still a value; only `undefined`
      // (i.e. resolveSystemPrompt returned `null`) means "omit".
      ...built.system !== void 0 ? { system: built.system } : {},
      messages: built.messages,
      ...config.body
    });
    const response = await fetch(url, {
      method: "POST",
      headers,
      body,
      signal
    });
    if (!response.ok) {
      throw await readErrorBody(response, PROVIDER_NAME);
    }
    for await (const payload of parseSSE(response, signal)) {
      let event;
      try {
        event = JSON.parse(payload);
      } catch (e) {
        continue;
      }
      if (event.type === "content_block_delta" && ((_d = event.delta) == null ? void 0 : _d.type) === "text_delta" && event.delta.text) {
        yield event.delta.text;
      }
      if (event.type === "message_stop") return;
    }
  };
}

export { createAnthropicProvider };
